File: gravitee-apim-repository/gravitee-apim-repository-api/src/main/java/io/gravitee/repository/management/model/ClientRegistrationProvider.java
/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.repository.management.model;

import java.util.Date;
import java.util.List;
import java.util.Objects;

public class ClientRegistrationProvider {

    public enum AuditEvent implements Audit.AuditEvent {
        CLIENT_REGISTRATION_PROVIDER_CREATED,
        CLIENT_REGISTRATION_PROVIDER_UPDATED,
        CLIENT_REGISTRATION_PROVIDER_DELETED,
    }

    /**
     * Client registration provider ID
     */
    private String id;

    /**
     * The ID of the environment the registration provider is attached to
     */
    private String environmentId;

    /**
     * Client registration provider name
     */
    private String name;

    /**
     * Client registration provider description
     */
    private String description;

    /**
     * Client registration OIDC discovery endpoint
     */
    private String discoveryEndpoint;

    private InitialAccessTokenType initialAccessTokenType;

    /**
     * Client registration OIDC Client_ID
     */
    private String clientId;

    /**
     * Client registration OIDC Client_secret
     */
    private String clientSecret;

    /**
     * Client registration OIDC scopes
     */
    private List<String> scopes;

    private String initialAccessToken;

    private boolean renewClientSecretSupport;

    private String renewClientSecretEndpoint;

    private String renewClientSecretMethod;

    private String softwareId;

    /**
     * Client registration provider creation date
     */
    private Date createdAt;

    /**
     * Client registration provider last updated date
     */
    private Date updatedAt;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getEnvironmentId() {
        return environmentId;
    }

    public void setEnvironmentId(String environmentId) {
        this.environmentId = environmentId;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(Date createdAt) {
        this.createdAt = createdAt;
    }

    public Date getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(Date updatedAt) {
        this.updatedAt = updatedAt;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getDiscoveryEndpoint() {
        return discoveryEndpoint;
    }

    public void setDiscoveryEndpoint(String discoveryEndpoint) {
        this.discoveryEndpoint = discoveryEndpoint;
    }

    public InitialAccessTokenType getInitialAccessTokenType() {
        return initialAccessTokenType;
    }

    public void setInitialAccessTokenType(InitialAccessTokenType initialAccessTokenType) {
        this.initialAccessTokenType = initialAccessTokenType;
    }

    public String getClientId() {
        return clientId;
    }

    public void setClientId(String clientId) {
        this.clientId = clientId;
    }

    public String getClientSecret() {
        return clientSecret;
    }

    public void setClientSecret(String clientSecret) {
        this.clientSecret = clientSecret;
    }

    public List<String> getScopes() {
        return scopes;
    }

    public void setScopes(List<String> scopes) {
        this.scopes = scopes;
    }

    public String getInitialAccessToken() {
        return initialAccessToken;
    }

    public void setInitialAccessToken(String initialAccessToken) {
        this.initialAccessToken = initialAccessToken;
    }

    public boolean isRenewClientSecretSupport() {
        return renewClientSecretSupport;
    }

    public void setRenewClientSecretSupport(boolean renewClientSecretSupport) {
        this.renewClientSecretSupport = renewClientSecretSupport;
    }

    public String getRenewClientSecretEndpoint() {
        return renewClientSecretEndpoint;
    }

    public void setRenewClientSecretEndpoint(String renewClientSecretEndpoint) {
        this.renewClientSecretEndpoint = renewClientSecretEndpoint;
    }

    public String getRenewClientSecretMethod() {
        return renewClientSecretMethod;
    }

    public void setRenewClientSecretMethod(String renewClientSecretMethod) {
        this.renewClientSecretMethod = renewClientSecretMethod;
    }

    public String getSoftwareId() {
        return softwareId;
    }

    public void setSoftwareId(String softwareId) {
        this.softwareId = softwareId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ClientRegistrationProvider that = (ClientRegistrationProvider) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    public enum InitialAccessTokenType {
        INITIAL_ACCESS_TOKEN,
        CLIENT_CREDENTIALS,
    }
}

File: gravitee-gamma/gravitee-gamma-module-platform/src/main/ui/features/client-registration/types/clientRegistrationProvider.ts
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export type InitialAccessTokenType = 'INITIAL_ACCESS_TOKEN' | 'CLIENT_CREDENTIALS';

export type RenewClientSecretMethod = 'POST' | 'PATCH' | 'PUT';

export type StoreType = 'NONE' | 'JKS' | 'PKCS12';

export interface CryptoStore {
    type: StoreType;
    password?: string;
    path?: string | null;
    content?: string | null;
    alias?: string;
    keyPassword?: string;
}

export interface ClientRegistrationProvider {
    id: string;
    name: string;
    description?: string;
    discovery_endpoint: string;
    initial_access_token_type: InitialAccessTokenType;
    client_id?: string;
    client_secret?: string;
    scopes?: string[];
    initial_access_token?: string;
    renew_client_secret_support?: boolean;
    renew_client_secret_endpoint?: string;
    renew_client_secret_method?: string;
    software_id?: string;
    updated_at?: number;
    claim_mappings?: Record<string, string>;
    trust_store?: CryptoStore;
    key_store?: CryptoStore;
}

/** POST /providers body — Classic `create(clientRegistrationProvider)` without an id. */
export interface ClientRegistrationProviderWrite {
    name: string;
    description?: string;
    discovery_endpoint: string;
    initial_access_token_type: InitialAccessTokenType;
    client_id?: string;
    client_secret?: string;
    scopes?: string[];
    initial_access_token?: string;
    renew_client_secret_support?: boolean;
    renew_client_secret_endpoint?: string;
    renew_client_secret_method?: string;
    software_id?: string;
    claim_mappings?: Record<string, string>;
    trust_store?: CryptoStore;
    key_store?: CryptoStore;
}

/**
 * PUT body field list from Classic `ClientRegistrationProvidersService.update`.
 * Do not send `id` in the JSON body.
 */
export const CLIENT_REGISTRATION_PROVIDER_UPDATE_FIELDS = [
    'name',
    'description',
    'discovery_endpoint',
    'initial_access_token_type',
    'client_id',
    'client_secret',
    'scopes',
    'initial_access_token',
    'renew_client_secret_support',
    'renew_client_secret_endpoint',
    'renew_client_secret_method',
    'software_id',
    'claim_mappings',
    'trust_store',
    'key_store',
] as const satisfies ReadonlyArray<keyof ClientRegistrationProviderWrite>;

export function toClientRegistrationProviderUpdateBody(provider: ClientRegistrationProviderWrite): ClientRegistrationProviderWrite {
    return {
        name: provider.name,
        description: provider.description,
        discovery_endpoint: provider.discovery_endpoint,
        initial_access_token_type: provider.initial_access_token_type,
        client_id: provider.client_id,
        client_secret: provider.client_secret,
        scopes: provider.scopes,
        initial_access_token: provider.initial_access_token,
        renew_client_secret_support: provider.renew_client_secret_support,
        renew_client_secret_endpoint: provider.renew_client_secret_endpoint,
        renew_client_secret_method: provider.renew_client_secret_method,
        software_id: provider.software_id,
        claim_mappings: provider.claim_mappings,
        trust_store: provider.trust_store,
        key_store: provider.key_store,
    };
}
